File: agent-service/src/core/agent-framework/registry.ts
import { BaseAgent } from "./base-agent";

type AgentFactory = (id: string, config?: Record<string, unknown>) => BaseAgent;

/**
 * Singleton registry that maps agent type strings to factory functions.
 *
 * Usage:
 *   AgentRegistry.getInstance().register("coder", (id, cfg) => new CoderAgent(id, cfg));
 *   const agent = AgentRegistry.getInstance().create("coder", "uuid-1");
 */
export class AgentRegistry {
  private static _instance: AgentRegistry | null = null;

  private factories = new Map<string, AgentFactory>();

  private constructor() {}

  static getInstance(): AgentRegistry {
    if (!AgentRegistry._instance) {
      AgentRegistry._instance = new AgentRegistry();
    }
    return AgentRegistry._instance;
  }

  /**
   * Register a factory for a given agent type.
   * Registering the same type twice overwrites the previous factory.
   */
  register(type: string, factory: AgentFactory): void {
    this.factories.set(type, factory);
  }

  /**
   * Create a new agent instance of the given type.
   * Throws if the type has not been registered.
   */
  create(type: string, id: string, config?: Record<string, unknown>): BaseAgent {
    const factory = this.factories.get(type);
    if (!factory) {
      throw new Error(
        `AgentRegistry: unknown type "${type}". Registered types: ${this.getRegisteredTypes().join(", ") || "(none)"}`,
      );
    }
    return factory(id, config);
  }

  /** Returns true if the given type has a registered factory. */
  has(type: string): boolean {
    return this.factories.has(type);
  }

  /** Returns the list of all registered agent type strings. */
  getRegisteredTypes(): string[] {
    return Array.from(this.factories.keys());
  }
}

File: agent-service/src/agent_service/core/agent_framework/registry.py
from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional

from agent_service.core.agent_framework.base_agent import BaseAgent

AgentFactory = Callable[[str, Optional[Dict[str, Any]]], BaseAgent]


class AgentRegistry:
    """
    Singleton registry that maps agent type strings to factory functions.

    Usage::

        AgentRegistry.get_instance().register("reviewer", lambda id, cfg: ReviewAgent(id))
        agent = AgentRegistry.get_instance().create("reviewer", "uuid-1")
    """

    _instance: Optional["AgentRegistry"] = None

    def __init__(self) -> None:
        self._factories: Dict[str, AgentFactory] = {}

    @classmethod
    def get_instance(cls) -> "AgentRegistry":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def register(self, type_: str, factory: AgentFactory) -> None:
        self._factories[type_] = factory

    def create(self, type_: str, id: str, config: Optional[Dict[str, Any]] = None) -> BaseAgent:
        factory = self._factories.get(type_)
        if factory is None:
            registered = ", ".join(self._factories.keys()) or "(none)"
            raise ValueError(
                f'AgentRegistry: unknown type "{type_}". Registered types: {registered}'
            )
        return factory(id, config)

    def has(self, type_: str) -> bool:
        return type_ in self._factories

    def get_registered_types(self) -> List[str]:
        return list(self._factories.keys())
